### PR TITLE
#14826: Remove misoptimizations from init code

### PR DESCRIPTION
**Pull request by nathan-TT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/14826

### Problem description
Empty kernels are larger than expected.

### What's changed
1) Stop wzerorange being recognized as memset. Memset is no longer pulled in.

2) Reduce insns in data image copy.  Original loop was 21 isnsn (3.5 per word), new loop is 10 insns (3.3 per word).

3) Do not use a loop for residue.  We only have to handle 0, 1 and 2 cases.  A loop is more overhead.

4) Sprinkle a few more unroll-inhibiting pragmas around.

Rename init code as do_crt1, to make it clearer what it is doing.

These changes remove 436 bytes from a kernel code.

### Checklist
- [ YES ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by nathan-TT** [comment]
I was dissatisfied with the volatile asms, so changed to actual assembly instructions, hiding the form of the loop from the compiler.

File: tt_metal/hw/toolchain/substitutes.cpp [hunk 1]
```diff
@@ -7,22 +7,18 @@
...
 }
 
-extern "C" void wzerorange(uint32_t *start, uint32_t *end) __attribute__((aligned(16)));
-
-extern "C" void wzerorange(uint32_t *start, uint32_t *end)
-{
-    for (; start != end; start++)
-    {
+extern "C" void wzerorange(uint32_t *start, uint32_t *end) {
+#pragma GCC unroll 0
```
[the remark this answers is not in the document]
Comment (nathan-TT) [reply tt_metal/hw/toolchain/substitutes.cpp:17]: wrong. `The values of @math{0} and @math{1} block any unrolling of the loop.`

File: tt_metal/hw/inc/firmware_common.h [hunk 2]
```diff
@@ -14,53 +14,56 @@
...
+    unsigned len = __ldm_data_end - __ldm_data_start;
+#pragma GCC unroll 0
+    while (len >= 3) {
+        auto v0 = src[0], v1 = src[1], v2 = src[2];
+        // Make sure optimizer does not think this is memcpy. The
+        // scheduler doesn't know the above loads have 6 cycle
+        // latency, so emit the 3 adds as a single block the optimizer
+        // will not be able to move. We don't need early clobbers here
+        // because of the +r constraint -- early clobbers would
+        // pessimize.
```
Comment (pgkeller) [inline tt_metal/hw/inc/firmware_common.h:41]: took me a moment, but I get it.  can you add to the comment something like: "loads have 6 cycle latency, do the pointer bookkeeping in the load shadow prior to the stores"

**Review by pgkeller — APPROVED** [review]
minor comment tweak requested

File: tt_metal/hw/inc/firmware_common.h (hunk 2 above)
Comment (nathan-TT) [reply tt_metal/hw/inc/firmware_common.h]: done

**Review by aliuTT — APPROVED** [review]

**Pull request merged**